File: 4_leetcode/C++/LC_1116_mut.cpp
#include <mutex>
#include <condition_variable>
#include <functional>

using namespace std;

class ZeroEvenOdd {
private:
    int n;
    int tick;
    int turn;
    mutex mtx;
    condition_variable cv;

public:
    ZeroEvenOdd(int n) : n(n), tick(1), turn(0) {}

    void zero(function<void(int)> printNumber) {
        unique_lock<mutex> lock(mtx);
        
        while (tick =n) {
            cv.wait(lock, [this] { return turn==0; });
            if (tick>n) break;
            
            printNumber(0);
            
            turn=(tick%2!=0) ? 1:2;
            cv.notify_all();
        }
    }

    void even(function<void(int)> printNumber) {
        unique_lock<mutex> lock(mtx);

        while (tick<=n) {
            cv.wait(lock, [this] { return turn==2; });
            if (tick>n) break;
            
            printNumber(tick++);
            
            turn=0;
            cv.notify_all();
        }
    }

    void odd(function<void(int)> printNumber) {
        unique_lock<mutex> lock(mtx);

        while (tick<=n) {
            cv.wait(lock, [this] { return turn==1; });
            if (tick>n) break;
            
            printNumber(tick++);
            
            turn=0;
            cv.notify_all();
        }
    }
};

File: 4_leetcode/C#/LC_1116_mut.cs
using System;
using System.Threading;

public class ZeroEvenOdd {
    private int n;

    private int zeroTick;
    private int count;
    private bool end;
    private object mut;
    
    public ZeroEvenOdd(int n) {
        this.n        =n;
        this.zeroTick =1;
        this.count    =1;
        this.end      =false;

        mut=new object();
    }

    // printNumber(x) outputs "x", where x is an integer.
    public void Zero(Action<int> printNumber) {

        while (true) {
            lock (mut) {
                while (zeroTick==0 && !end) Monitor.Wait(mut);

                if (end) return;

                printNumber(0);
                zeroTick=0;
                Monitor.PulseAll(mut);
            }
        }
    }

    public void Even(Action<int> printNumber) {
        
        while (true) {
            lock (mut) {
                while ((zeroTick==1||count%2!=0) && !end) {
                    Monitor.Wait(mut);
                }

                if (end) return;

                printNumber(count);
                count++;
                zeroTick=1;

                if (count>n) {
                    end=true;
                    Monitor.PulseAll(mut);
                    return;
                }

                Monitor.PulseAll(mut);
            }
        }
    }

    public void Odd(Action<int> printNumber) {
        while (true) {
            lock (mut) {
                while ((zeroTick==1||count%2==0) && !end) {
                    Monitor.Wait(mut);
                }

                if (end) return;

                printNumber(count);
                count++;
                zeroTick=1;

                if (count>n) {
                    end=true;
                    Monitor.PulseAll(mut);
                    return;
                }

                Monitor.PulseAll(mut);
            }
        }
    }
}